apps/rush-lib/src/logic/deploy/DeployManager.ts

```diff
@@ -0,0 +1,588 @@
+
+  /**
+   * The absolute path of the target folder for the subdeployment. If enableSubdeployments=false,
+   * then this points to the DeployManager._targetRootFolder.
+   */
+  targetSubdeploymentFolder: string;
+
+  /**
+   * During the analysis stage, _collectFoldersRecursive() uses this set to collect the absolute paths
+   * of the package folders to be copied.  The copying is performed later by _deployFolder().
+   */
+  foldersToCopy: Set<string>;
+
+  /**
+   * Additional information about some of the foldersToCopy paths.
+   * The key is the absolute real path from foldersToCopy.
+   */
+  folderInfosByPath: Map<string, IFolderInfo>;
+
+  symlinkAnalyzer: SymlinkAnalyzer;
+}
+
+/**
+ * Manages the business logic for the "rush deploy" command.
+ */
+export class DeployManager {
+  private static _jsonSchema: JsonSchema = JsonSchema.fromFile(
+    path.join(__dirname, '../../schemas/deploy-scenario.schema.json')
+  );
+
+  // Used by validateScenarioName()
+  private static _scenarioNameRegExp: RegExp = /^[a-z0-9]+(-[a-z0-9]+)*$/;
+
+  private readonly _rushConfiguration: RushConfiguration;
+  private readonly _packageJsonLookup: PackageJsonLookup;
+
+  /**
+   * The target folder for the deployment.  By default it will be "common/deploy".
+   */
+  private _targetRootFolder: string;
+  /**
+   * The source folder that copying originates from.  Generally it is the repo root folder with rush.json.
+   */
+  private _sourceRootFolder: string;
+
+  /**
+   * The parsed scenario config file, as defined by the "deploy-scenario.schema.json" JSON schema
+   */
+  private _deployScenarioJson: IDeployScenarioJson;
+
+  /**
+   * Used to lookup items in IDeployScenarioJson.projectSettings based on their IDeployScenarioProjectJson.projectName
+   */
+  private _deployScenarioProjectJsonsByName: Map<string, IDeployScenarioProjectJson>;
+
+  public constructor(rushConfiguration: RushConfiguration) {
+    this._rushConfiguration = rushConfiguration;
+    this._packageJsonLookup = new PackageJsonLookup();
+    this._deployScenarioProjectJsonsByName = new Map();
+  }
+
+  /**
+   * Validates that the input string conforms to the naming rules for a "rush deploy" scenario name.
+   */
+  public static validateScenarioName(scenarioName: string): void {
+    if (!scenarioName) {
+      throw new Error('The scenario name cannot be an empty string');
+    }
+    if (!this._scenarioNameRegExp.test(scenarioName)) {
+      throw new Error(`"${scenarioName}" is not a valid scenario name. The name must be comprised of`
+        + ' lower case letters and numbers, separated by single hyphens. Example: "my-scenario"');
+    }
+  }
+
+  /**
+   * Load and validate the scenario config file.  The result is stored in this._deployScenarioJson.
+   */
+  private _loadConfigFile(scenarioName: string): void {
+    const deployScenarioPath: string = path.join(this._rushConfiguration.commonFolder, 'config/deploy-scenarios',
+      scenarioName + '.json');
```

Comment:
This same statement exists in the action. Dedupe them by sticking it in a utility somewhere?

Reply:
Fixed